source/base/mpi.cc

```diff
@@ -812,8 +814,468 @@ namespace Utilities
+              is.compress();
+
+              // translate index set to a list of pairs
+              auto &buffer = buffers[index];
+              for (auto interval = is.begin_intervals();
+                   interval != is.end_intervals();
+                   interval++)
+                buffer.emplace_back(*interval->begin(), interval->last() + 1);
+
+              // send data
+              MPI_Isend(&buffer[0],
+                        buffer.size() * 2,
+                        DEAL_II_DOF_INDEX_MPI_TYPE,
+                        rank,
+                        StateMachine::tag_setup,
+                        comm,
+                        &request[index]);
+            }
+        }
+
+
+        // 4) receive messages until all dofs in dict are processed
+        while (dict.local_size != dic_local_rececived)
+          {
+            // wait for an incoming massage
+            MPI_Status status;
+            MPI_Probe(MPI_ANY_SOURCE, StateMachine::tag_setup, comm, &status);
+
+            // retrieve size of incoming massage
+            int number_amount;
+            MPI_Get_count(&status, DEAL_II_DOF_INDEX_MPI_TYPE, &number_amount);
+
+            auto other_rank = status.MPI_SOURCE;
+
+            // receive massage
+            std::vector<
+              std::pair<types::global_dof_index, types::global_dof_index>>
+              buffer(number_amount / 2);
+            MPI_Recv(&buffer[0],
+                     number_amount,
+                     DEAL_II_DOF_INDEX_MPI_TYPE,
+                     other_rank,
+                     StateMachine::tag_setup,
+                     comm,
+                     &status);
+
+            // process message: loop over all intervals
+            for (auto interval : buffer)
+              for (types::global_dof_index i = interval.first;
+                   i < interval.second;
+                   i++)
+                {
+                  dict.array[i - dict.local_range.first] = other_rank;
+                  dic_local_rececived++;
+                }
+          }
+
+        // 5) make sure that all messages have been sent
+        std::vector<MPI_Status> status(relevant_procs_size);
+        MPI_Waitall(relevant_procs_size, &request[0], &status[0]);
+      }
+
+
+    } // namespace ComputeIndexOwner
+
+#endif
+
+    std::vector<unsigned int>
+    compute_index_owner(const IndexSet &owned_dofs,
+                        const IndexSet &ghosted_dofs,
+                        const MPI_Comm &comm)
+    {
+#ifdef DEAL_II_WITH_MPI
+      using namespace ComputeIndexOwner;
+
+      StateMachine state_machine(comm);
+
+      // Step 1: setup dictionary
+      setup_dictionary(state_machine.dict, owned_dofs, comm);
+
+      // Step 2: read dictionary
+      std::vector<unsigned int> array(ghosted_dofs.n_elements());
+      state_machine.run(array, ghosted_dofs);
+
+      return array;
+#else
+      // if compiled without MPI: there are no neighbors...
+      (void)owned_dofs;
```

Comment:
Can you assert that ghosted_dofs is a subset of owned_dofs?

Reply:
@tjhei: How can I do this? Is it enough to:
```cpp
Assert(owned_dofs.size() == ghosted_dofs.size(),
             ExcMessage("IndexSets have to have the same sizes."));
```

Reply:
I think what he meant is that `ghosted_dofs` is a subset of the *union* of the `owned_dofs`.